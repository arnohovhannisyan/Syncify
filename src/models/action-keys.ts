import { ActionTypes } from "~/redux/actions";

export type ActionKeys = keyof typeof ActionTypes;
