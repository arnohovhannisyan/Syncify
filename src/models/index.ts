export * from "~/models/syncers";
export * from "~/models/languagePack";
export * from "~/models/profile";
export * from "~/models/migrations";
export * from "~/models/settings";
export * from "~/models/syncer";
export * from "~/models/webview/settingType";
export * from "~/models/webview/section";
export * from "~/models/webview/setting";
export * from "~/models/webview/selectOption";
