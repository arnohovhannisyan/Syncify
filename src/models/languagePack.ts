export type LanguagePack = {
	[key: string]: string | undefined;
};
