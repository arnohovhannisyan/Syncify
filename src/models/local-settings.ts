export type LocalSettings = {
	/**
	 * The method used to sync your settings.
	 */
	syncer: "git" | "file";

	/**
	 * The profile currently being used.
	 */
	currentProfile: string;

	/**
	 * The path where you want your settings to be saved to.
	 *
	 * - For the `git` syncer, this is the URI of the repository.
	 *
	 * - For the `file` syncer, this is the URI of your export folder.
	 */
	exportPath: string;

	/**
	 * Files that will be uploaded, formatted as an array of globs. If a path matches a listed glob, it will be uploaded.
	 */
	filesToInclude: string[];

	/**
	 * Hostnames are used by `Sync Pragmas` to differentiate between different computers.
	 */
	hostname: string;

	/**
	 * Controls whether or not local settings will be forcefully uploaded, even if remote settings are up to date or there are newer remote changes.
	 */
	forceUpload: boolean;

	/**
	 * Controls whether or not remote settings will be forcefully downloaded, even if local settings are up to date.
	 */
	forceDownload: boolean;
};

export const defaultLocalSettings: LocalSettings = {
	syncer: "git",
	currentProfile: "main",
	exportPath: "",
	filesToInclude: ["**/*", "!workspaceStorage", "!globalStorage"],
	hostname: "",
	forceDownload: false,
	forceUpload: false,
};
