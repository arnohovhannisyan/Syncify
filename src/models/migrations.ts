export type Migration = (previousVersion: string) => void | Promise<void>;

export interface Migrations {
	[key: string]: Migration;
}
