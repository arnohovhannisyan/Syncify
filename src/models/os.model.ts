export enum OperatingSystem {
  Windows = "win32",
  Mac = "darwin",
  Linux = "linux"
}
