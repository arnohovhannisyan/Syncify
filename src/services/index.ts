export * from "~/services/environment.service";
export * from "~/services/localization.service";
export * from "~/services/custom-file.service";
export * from "~/services/extensions.service";
export * from "~/services/factory.service";
export * from "~/services/fs.service";
export * from "~/services/oauth.service";
export * from "~/services/init.service";
export * from "~/services/logger.service";
export * from "~/services/pragma.service";
export * from "~/services/settings.service";
export * from "~/services/utilities.service";
export * from "~/services/watcher.service";
export * from "~/services/webview.service";
