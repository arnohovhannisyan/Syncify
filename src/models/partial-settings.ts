import { DeepPartial } from "utility-types";
import { ISettings } from "~/models";

export type PartialSettings = DeepPartial<ISettings>;
