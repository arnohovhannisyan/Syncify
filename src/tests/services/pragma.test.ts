import { OperatingSystem } from "~/models";
import { Environment, Pragma } from "~/services";

jest.mock("~/services/localization.ts");

it("should properly process before uploading", () => {
  const initial = `{
    // @sync host=jest
    // "abc": "xyz"
  }`;

  const expected = `{
    // @sync host=jest
    "abc": "xyz"
  }`;

  expect(Pragma.processOutgoing(initial)).toBe(expected);
});

it("should return new content if invalid", () => {
  const newContent = `Invalid JSON`;

  expect(Pragma.processIncoming("", newContent, "")).toBe(newContent);
});

it("should properly ignore", () => {
  const initial = `{
    // @sync-ignore
    "abc": "xyz",
    // @sync-ignore
    "abc2": "xyz2",
    "xyz": "abc"
  }`;

  const expected = `{
    "xyz": "abc"
  }`;

  expect(Pragma.processOutgoing(initial)).toBe(expected);

  const initial2 = `{
    "xyz": "def"
  }`;

  const expected2 = `{
    // @sync-ignore
    "abc": "xyz",
    // @sync-ignore
    "abc2": "xyz2",


    "xyz": "def"
  }`;

  expect(Pragma.processIncoming("", initial2, initial)).toBe(expected2);
});

it("should properly handle brackets", () => {
  const initial = `{
    // @sync host=jest
    // "abc": "{xyz}"
  }`;

  const expected = `{
    // @sync host=jest
    "abc": "{xyz}"
  }`;

  expect(Pragma.processOutgoing(initial)).toBe(expected);

  const initial2 = `{
    // @sync host=jest
    "abc": "[xyz]"
  }`;

  const expected2 = `{
    // @sync host=jest
    // "abc": "[xyz]"
  }`;

  expect(Pragma.processIncoming("test", initial2, initial2)).toBe(expected2);
});

describe("host", () => {
  it("should work with 'host'", () => {
    const input = `{
      // @sync host=jest
      // "abc": "xyz"
    }`;

    const withCorrectHost = `{
      // @sync host=jest
      "abc": "xyz"
    }`;

    const withIncorrectHost = `{
      // @sync host=jest
      // "abc": "xyz"
    }`;

    expect(Pragma.processIncoming("test", input)).toBe(withIncorrectHost);
    expect(Pragma.processIncoming("jest", input)).toBe(withCorrectHost);
  });
});

describe("os", () => {
  Object.keys(OperatingSystem).forEach(key => {
    it(`should work on 'OperatingSystem.${key}'`, () => {
      const os = OperatingSystem[key as keyof typeof OperatingSystem];

      const input = `{
        // @sync os=${key.toLowerCase()}
        // "abc": "xyz"
      }`;

      const withCorrectOS = `{
        // @sync os=${key.toLowerCase()}
        "abc": "xyz"
      }`;

      const withIncorrectOS = `{
        // @sync os=${key.toLowerCase()}
        // "abc": "xyz"
      }`;

      Environment.os = os;

      expect(Pragma.processIncoming("", input)).toBe(withCorrectOS);

      Environment.os =
        os === OperatingSystem.Windows
          ? OperatingSystem.Mac
          : OperatingSystem.Windows;

      expect(Pragma.processIncoming("", input)).toBe(withIncorrectOS);
    });
  });
});

describe("env", () => {
  it("should work with 'env'", () => {
    const input = `{
      // @sync env=SYNCIFY
      "abc": "xyz"
    }`;

    const withCorrectEnv = `{
      // @sync env=SYNCIFY
      "abc": "xyz"
    }`;

    const withIncorrectEnv = `{
      // @sync env=SYNCIFY
      // "abc": "xyz"
    }`;

    process.env.SYNCIFY = "true";

    expect(Pragma.processIncoming("", input)).toBe(withCorrectEnv);

    process.env.SYNCIFY = "";

    expect(Pragma.processIncoming("", input)).toBe(withIncorrectEnv);
  });
});
