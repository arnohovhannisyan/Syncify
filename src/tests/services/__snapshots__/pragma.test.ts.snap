// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`incoming env 1`] = `
"{
				// @sync env=ENV_TEST_KEY
				// \\"key\\": \\"value\\"
			}"
`;

exports[`incoming env 2`] = `
"{
				// @sync env=ENV_TEST_KEY
				\\"key\\": \\"value\\"
			}"
`;

exports[`incoming host 1`] = `
"{
				// @sync host=pc
				\\"key\\": \\"value\\"
			}"
`;

exports[`incoming host 2`] = `
"{
				// @sync host=invalid
				// \\"key\\": \\"value\\"
			}"
`;

exports[`incoming operating system 1`] = `
"{
				// @sync os=windows
				\\"key\\": \\"value\\"
			}"
`;

exports[`incoming operating system 2`] = `
"{
				// @sync os=mac
				// \\"key\\": \\"value\\"
			}"
`;

exports[`incoming operating system 3`] = `
"{
				// @sync os=linux
				// \\"key\\": \\"value\\"
			}"
`;

exports[`outgoing 1`] = `
"{
		// @sync ...
		// \\"key\\": \\"value\\",

		// @invalid
		\\"key2\\": \\"value2\\"
	}"
`;
